scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -0,0 +1,862 @@
+                }
+                break;
+            case Level.INFO_INT:
+                if (logger.isInfoEnabled()) {
+                    TaskLogger.getInstance().info(taskId, channelHeader + message);
+                }
+                break;
+            case Level.WARN_INT:
+                TaskLogger.getInstance().warn(taskId, channelHeader + message);
+                break;
+            case Level.ERROR_INT:
+                TaskLogger.getInstance().error(taskId, channelHeader + message);
+                break;
+            default:
+                // no action
+        }
+        if (logger.isTraceEnabled() && channel != null) {
+            TaskLogger.getInstance().info(taskId, channelHeader + "New channel content : " + getChannel(channel));
+        }
+    }
+
+    @Override
+    public void runActivity(Body body) {
+        Service service = new Service(body);
+        while (body.isActive()) {
+            try {
+                Request request = service.blockingRemoveOldest();
+                if (request == null) {
+                    return;
+                }
+                if (request.getMethodName().startsWith("wait") && !testWaitFunction(service, request)) {
+                    waitUntilQueue.add(request);
+                } else {
+                    service.serve(request);
+                    unblockWaitMethods(service);
+                }
+
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    /**
+     * Unblock all pending wait methods if predicates are met
+     * @param service active object service
+     */
+    private void unblockWaitMethods(Service service) {
+        for (Iterator<Request> iterator = waitUntilQueue.iterator(); iterator.hasNext();) {
+            Request waitRequest = iterator.next();
+            if (testWaitFunction(service, waitRequest)) {
+                service.serve(waitRequest);
+                iterator.remove();
+            }
+        }
+    }
+
+    private <T> T evaluateClosure(String closureDefinition, Class<T> type) throws CompilationException {
+        try {
+            return (T) shell.evaluate(closureDefinition + " as " + type.getCanonicalName());
+        } catch (CompilationFailedException e) {
+            // CompilationFailedException contains instances which are not serializable
+            throw new CompilationException(StackTraceUtil.getStackTrace(e));
+        }
+    }
+
+    /**
+     * Execute the waitUntil function corresponding to the provided request
+     *
+     * If an error occurs, serve the request to unblock the caller with the exception attached
+     *
+     * @param service active object service
+     * @param request request to process
+     * @return true if the result of the waitUntil call is true or if an exception occurred
+     */
+    private boolean testWaitFunction(Service service, Request request) {
```

Comment:
This makes me think: in the Synchronization API with did not provide a flavor of the `wait` methods with a timeout option. Would it make sense?

Reply:
It would make sense yes.